xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp

```diff
@@ -0,0 +1,444 @@
+  while ((pyIterResult = PyIter_Next(pyResultIterator)) != NULL)
+  {
+    std::string result;
+    try
+    {
+      PythonBindings::PyXBMCGetUnicodeString(result, pyIterResult, false, "result", "handle_request");
+    }
+    catch (const XBMCAddon::WrongTypeException& e)
+    {
+      CLog::Log(LOGERROR, "CHTTPPythonWsgiInvoker: failed to parse result iterable object for WSGI script \"%s\" with wrong type exception: %s", script.c_str(), e.GetMessage());
+      goto cleanup;
+    }
+    catch (const XbmcCommons::Exception& e)
+    {
+      CLog::Log(LOGERROR, "CHTTPPythonWsgiInvoker: failed to parse result iterable object for WSGI script \"%s\" with exception: %s", script.c_str(), e.GetMessage());
+      goto cleanup;
+    }
+    catch (...)
+    {
+      CLog::Log(LOGERROR, "CHTTPPythonWsgiInvoker: failed to parse result iterable object for WSGI script \"%s\" with unknown exception", script.c_str());
+      goto cleanup;
+    }
+
+    // append the result string to the response
+    m_wsgiResponse->Append(result);
+  }
+
+cleanup:
+  if (pyIterResult != NULL)
+    Py_DECREF(pyIterResult);
+  if (pyResultIterator != NULL)
+  {
+    // Call optional close method on iterator
+    if (PyObject_HasAttrString(pyResultIterator, (char*)"close") == 1)
+    {
+      if (PyObject_CallMethod(pyResultIterator, (char*)"close", NULL) == NULL)
+        CLog::Log(LOGERROR, "CHTTPPythonWsgiInvoker: failed to close iterator object for WSGI script \"%s\"", script.c_str());
+    }
+    Py_DECREF(pyResultIterator);
+  }
+  if (pyResult != NULL)
+    Py_DECREF(pyResult);
+  if (pyEntryPoint != NULL)
+    Py_DECREF(pyEntryPoint);
+  if (pyModule != NULL)
+    Py_DECREF(pyModule);
+}
+
+std::map<std::string, CPythonInvoker::PythonModuleInitialization> CHTTPPythonWsgiInvoker::getModules() const
+{
+  static std::map<std::string, PythonModuleInitialization> modules;
+  if (modules.empty())
+  {
+    for (size_t i = 0; i < PythonModulesSize; i++)
+      modules.insert(std::make_pair(PythonModules[i].name, PythonModules[i].initialization));
+  }
+
+  return modules;
+}
+
+const char* CHTTPPythonWsgiInvoker::getInitializationScript() const
+{
+  return RUNSCRIPT;
+}
+
+std::map<std::string, std::string> CHTTPPythonWsgiInvoker::createCgiEnvironment(const HTTPPythonRequest* httpRequest)
+{
+  std::map<std::string, std::string> environment;
+
+  // REQUEST_METHOD
+  std::string requestMethod;
+  switch (httpRequest->method)
+  {
+  case HEAD:
+    requestMethod = "HEAD";
+    break;
+
+  case POST:
+    requestMethod = "POST";
+    break;
+
+  case GET:
+  default:
+    requestMethod = "GET";
+    break;
+  }
+  environment.insert(std::make_pair("REQUEST_METHOD", requestMethod));
+
+  // SCRIPT_NAME
+  std::string path = httpRequest->path;
+  if (path.compare("/") == 0)
+    path.clear();
+  environment.insert(std::make_pair("SCRIPT_NAME", path));
+
+  // TODO: PATH_INFO
+  environment.insert(std::make_pair("PATH_INFO", path));
```

Comment:
IMO the specification of PATH_INFO is anything but obvious. I've looked through the web and found several questions concerning it and often the answer is something like "I think [...]".